plugins.py

```diff
@@ -0,0 +1,83 @@
+import time
+import datetime
+import deltachat
+
+
+class ReceivePlugin:
+    name = "test account"
+
+    @staticmethod
+    @deltachat.account_hookimpl
+    def ac_incoming_message(message):
+        received = time.time()
+        message.create_chat()
+        if not message.chat.can_send():
+            # if it's a device message or mailing list, we don't need to look at it.
+            return
+        sender = message.get_sender_contact().addr
+        receiver = message.account.get_self_contact().addr
+        msgcontent = parse_msg(message.text)
+        firsttravel = msgcontent.get("testduration")
+        secondtravel = received - msgcontent.get("begin")
+        iam = msgcontent.get("sender")
+        if message.chat.is_group():
+            if iam == "spider":
+                print("%s: joined group chat %s after %.1f seconds" % (receiver, message.chat.get_name(), secondtravel))
+                message.chat.send_text("Sender: %s\nBegin: %s" % (receiver, str(time.time())))
+                return  # for now it's enough to just accept the group chat
+            else:
+                return  # when messages by others arrived is checked later by the main thread
+        print("%s: test message took %.1f seconds to %s and %.1f seconds back." %
+              (receiver, firsttravel, sender, secondtravel))
+        message.account.output.submit_1on1_result(receiver, firsttravel, secondtravel)
+        message.account.shutdown()
+
+
+class EchoPlugin:
+    name = "spider"
+
+    @staticmethod
+    @deltachat.account_hookimpl
+    def ac_incoming_message(message):
+        received = time.time()
+        message.create_chat()
+        addr = message.get_sender_contact().addr
+        if message.is_system_message():
+            message.chat.send_text("echoing system message from {}:\n{}".format(addr, message))
+        elif message.chat.is_group():
+            return  # can safely ignore group messages. spider only creates it
+        else:
+            sent = parse_msg(message.text).get("begin")
+            testduration = received - sent
+            msginfo = message.get_message_info()
+            begin = time.time()
+            message.chat.send_text("TestDuration: %f\nBegin: %f\n%s" % (testduration, begin, msginfo))
```

Comment:
The spider handles incoming messages differently than the test accounts of course (it also gets different messages). But this method grew organically, the same way as the test account's method, and it shows in the messy structure. This could use some refactoring as well.

Reply:
The spider receives at least:

- a single message (soon with a large file, see #8)
- all of the group responses, which it doesn't need to handle